File: calculation-engine/csharp/src/CalculationEngine/Expression.cs
namespace CalculationEngine;

public sealed class Expression
{
    private static readonly int _precmin;
    private static readonly int[] _precedence;

    internal readonly int _requiredInputCount;
    internal readonly List<Node> _expression;

    private static int Prec(Node n) => n is Operator op ? _precedence[(int)op.Value - _precmin] : -1;

    static Expression()
    {
        var ops = new[] { (Op: '+', P: 1), (Op: '-', P: 1), (Op: '*', P: 2), (Op: '/', P: 2), (Op: '^', P: 3), };
        _precmin = ops.Min(o => (int)o.Op);
        var max = ops.Max(o => (int)o.Op);
        _precedence = new int[max - _precmin + 1];
        for (int i = 0; i < max - _precmin + 1; i++)
        {
            (char Op, int P)? op = null;
            for (int j = 0; j < ops.Length; j++)
            {
                if ((int)ops[j].Op - _precmin == i)
                {
                    op = ops[j];
                    break;
                }
            }

            if (op is var (ch, p))
            {
                _precedence[i] = p;
            }
        }
    }

    private Expression(List<Node> expression, int requiredInputCount)
    {
        _expression = expression;
        _requiredInputCount = requiredInputCount;
    }

    public ScalarEngine ToScalarEngine() => new ScalarEngine(this);

    public VectorizedEngine ToVectorizedEngine() => new VectorizedEngine(this);

    public static Expression FromInfix(IReadOnlyList<Node> expression)
    {
        if (expression is null || expression.Count == 0)
            ThrowHelper.ThrowArgumentException("Invalid expression");

        var result = new List<Node>(expression.Count);
        var stack = new Stack<Node>();

        for (int i = 0; i < expression.Count; i++)
        {
            var op = expression[i];

            if (op is Operand)
                result.Add(op);
            else if (op is LeftParens)
                stack.Push(op);
            else if (op is RightParens)
            {
                while (stack.Count > 0 && stack.Peek() is not LeftParens)
                    result.Add(stack.Pop());

                if (stack.Count > 0 && stack.Peek() is not LeftParens)
                    ThrowHelper.ThrowArgumentException("Invalid expression");

                stack.Pop();
            }
            else
            {
                var prec = Prec(op);
                while (stack.Count > 0 && prec <= Prec(stack.Peek()))
                    result.Add(stack.Pop());

                stack.Push(op);
            }
        }

        while (stack.TryPop(out var op))
            result.Add(op);

        return new Expression(result, result.Count(n => n is Operand));
    }
}

public abstract record Node { }

public sealed record Operand : Node { }

public sealed record LeftParens : Node { }

public sealed record RightParens : Node { }

public sealed record Operator : Node
{
    public static readonly Operator Add = new Operator('+');
    public static readonly Operator Sub = new Operator('-');
    public static readonly Operator Mul = new Operator('*');
    public static readonly Operator Div = new Operator('/');

    public char Value { get; }

    private Operator(char value)
    {
        Value = value;
    }
}

File: calculation-engine/go/pkg/expression.go
package expression

import (
	"calcengine/internal"
	"errors"
)

type Expression struct {
	expression         []Node
	requiredInputCount int
}

func FromInfix(expression []Node) (*Expression, error) {
	if len(expression) == 0 {
		return nil, errors.New("expression cannot be nil or empty")
	}

	result := make([]Node, 0, len(expression))
	stack := internal.NewStack[Node]()

	for _, op := range expression {

		nodeType := op.Type()
		if nodeType == OperandType {
			result = append(result, op)
		} else if nodeType == LeftParensType {
			stack.Push(op)
		} else if nodeType == RightParensType {
			for n := stack.Peek(); n != nil && (*n).Type() != LeftParensType; {
				result = append(result, *stack.Pop())
			}

			if n := stack.Peek(); n != nil && (*n).Type() != LeftParensType {
				return nil, errors.New("invalid expression")
			}

			stack.Pop()
		} else {
			prec := precedence(op)
			for n := stack.Peek(); n != nil && prec <= precedence(*n); {
				result = append(result, *stack.Pop())
			}

			stack.Push(op)
		}
	}

	for stack.Len() > 0 {
		result = append(result, *stack.Pop())
	}

	requiredInputCount := 0
	for _, n := range result {
		if n.Type() == OperandType {
			requiredInputCount++
		}
	}

	return &Expression{
		expression:         result,
		requiredInputCount: requiredInputCount,
	}, nil
}

func precedence(n Node) int {
	nodeType := n.Type()
	if nodeType != OperatorType {
		return -1
	}

	switch n.(*Operator).value {
	case '+':
	case '-':
		return 1
	case '*':
	case '/':
		return 2
	default:
		return -1
	}
	return -1
}

type NodeType int

const (
	OperandType     = iota
	LeftParensType  = iota
	RightParensType = iota
	OperatorType    = iota
)

type Node interface {
	Type() NodeType
}

type Operand struct {
}

func (o *Operand) Type() NodeType {
	return OperandType
}

type LeftParens struct {
}

func (o *LeftParens) Type() NodeType {
	return LeftParensType
}

type RightParens struct {
}

func (o *RightParens) Type() NodeType {
	return RightParensType
}

var Add Operator = Operator{value: '+'}
var Sub Operator = Operator{value: '-'}
var Mul Operator = Operator{value: '*'}
var Div Operator = Operator{value: '/'}

type Operator struct {
	value rune
}

func (o *Operator) Type() NodeType {
	return OperatorType
}
